File: api/internal/utils/sms/sms_of_aliyun.go
package sms

import (
	"context"
	"errors"
	"strings"

	openapi "github.com/alibabacloud-go/darabonba-openapi/v2/client"
	dysmsapi20170525 "github.com/alibabacloud-go/dysmsapi-20170525/v3/client"
	util "github.com/alibabacloud-go/tea-utils/v2/service"
	"github.com/alibabacloud-go/tea/tea"
	"github.com/gogf/gf/v2/util/gconv"
)

type SmsOfAliyun struct {
	AccessKeyId     string `json:"accessKeyId"`
	AccessKeySecret string `json:"accessKeySecret"`
	Endpoint        string `json:"endpoint"`
	SignName        string `json:"signName"`
	TemplateCode    string `json:"templateCode"`
}

func NewSmsOfAliyun(ctx context.Context, config map[string]any) *SmsOfAliyun {
	smsObj := &SmsOfAliyun{}
	gconv.Struct(config, smsObj)
	if smsObj.AccessKeyId == `` || smsObj.AccessKeySecret == `` || smsObj.Endpoint == `` {
		panic(`缺少插件配置：短信-阿里云`)
	}
	/* if smsObj.SignName == `` || smsObj.TemplateCode == `` {
		panic(`缺少插件配置：短信-阿里云短信模板`)
	} */
	return smsObj
}

func (smsThis *SmsOfAliyun) SendCode(ctx context.Context, phone string, code string) (err error) {
	if smsThis.SignName == `` || smsThis.TemplateCode == `` {
		err = errors.New(`缺少插件配置：短信-阿里云短信模板`)
		return
	}
	err = smsThis.SendSms(ctx, []string{phone}, `{"code": "`+code+`"}`, smsThis.SignName, smsThis.TemplateCode)
	return
}

func (smsThis *SmsOfAliyun) SendSms(ctx context.Context, phoneArr []string, message string, paramOpt ...any) (err error) {
	if len(paramOpt) < 2 {
		err = errors.New(`缺少签名和模板参数`)
		return
	}

	client, err := smsThis.CreateClient()
	if err != nil {
		return
	}

	sendSmsRequest := &dysmsapi20170525.SendSmsRequest{
		PhoneNumbers:  tea.String(strings.Join(phoneArr, `,`)),
		SignName:      tea.String(gconv.String(paramOpt[0])),
		TemplateCode:  tea.String(gconv.String(paramOpt[1])),
		TemplateParam: tea.String(message),
	}

	tryErr := func() (_e error) {
		defer func() {
			r := tea.Recover(recover())
			if r != nil {
				_e = r
			}
		}()
		result, err := client.SendSmsWithOptions(sendSmsRequest, &util.RuntimeOptions{})
		if err != nil {
			return err
		}
		if *result.Body.Code != `OK` {
			err = errors.New(*result.Body.Message)
			return err
		}
		return nil
	}()

	if tryErr != nil {
		var errSDK = &tea.SDKError{}
		if _t, ok := tryErr.(*tea.SDKError); ok {
			errSDK = _t
		} else {
			errSDK.Message = tea.String(tryErr.Error())
		}
		errMsg, errSms := util.AssertAsString(errSDK.Message)
		if errSms != nil {
			err = errSms
			return
		}
		err = errors.New(*errMsg)
	}
	return
}

func (smsThis *SmsOfAliyun) CreateClient() (client *dysmsapi20170525.Client, err error) {
	config := &openapi.Config{
		AccessKeyId:     tea.String(smsThis.AccessKeyId),
		AccessKeySecret: tea.String(smsThis.AccessKeySecret),
		Endpoint:        tea.String(smsThis.Endpoint),
	}
	client, err = dysmsapi20170525.NewClient(config)
	return
}

File: api_sw/app/Plugin/Sms/SmsOfAliyun.php
<?php

declare(strict_types=1);

namespace App\Plugin\Sms;

use AlibabaCloud\SDK\Dysmsapi\V20170525\Dysmsapi;
use \Exception;
use AlibabaCloud\Tea\Exception\TeaError;
use AlibabaCloud\Tea\Utils\Utils;
use Darabonba\OpenApi\Models\Config;
use AlibabaCloud\SDK\Dysmsapi\V20170525\Models\SendSmsRequest;
use AlibabaCloud\Tea\Utils\Utils\RuntimeOptions;

class SmsOfAliyun extends AbstractSms
{
    protected $config = [];

    /**
     * 发送短信
     *
     * @param string $phone
     * @param string $code
     * @return void
     */
    public function send(string $phone, string $code)
    {
        return $this->sendSms([$phone], json_encode(['code' => $code], JSON_UNESCAPED_UNICODE));
    }

    /**
     * 发送短信(批量)
     *
     * @param array $phoneArr
     * @param string $templateParam
     * @return void
     */
    public function sendSms(array $phoneArr, string $templateParam)
    {
        $client = $this->createClient();
        $sendSmsRequest = new SendSmsRequest([
            'phoneNumbers' => implode(',', $phoneArr),
            'signName' => $this->config['signName'],
            'templateCode' => $this->config['templateCode'],
            'templateParam' => $templateParam,
            // 'templateParam' => '{"code": "1234"}',
        ]);
        try {
            $result = $client->sendSmsWithOptions($sendSmsRequest, new RuntimeOptions([]));
            $result = $result->toMap();
        } catch (Exception $error) {
            if (!($error instanceof TeaError)) {
                $error = new TeaError([], $error->getMessage(), $error->getCode(), $error);
            }
            $errMsg = Utils::assertAsString($error->message);
            if (!empty($errMsg)) {
                throwFailJson(79999999, $errMsg);
            }
        }
        if (!(isset($result['body']['Code']) && $result['body']['Code'] == 'OK')) {
            throwFailJson(79999999, $result['body']['Message'] ?? '');
        }
    }

    /**
     * 使用AK&SK初始化账号Client
     * @param string $accessKeyId
     * @param string $accessKeySecret
     * @return Dysmsapi Client
     */
    public function createClient()
    {
        $config = new Config([
            'accessKeyId' => $this->config['accessKeyId'],
            'accessKeySecret' => $this->config['accessKeySecret'],
            'endpoint' => $this->config['endpoint']
        ]);
        return new Dysmsapi($config);
    }
}
